<span style="color:#b30000;font-size:25px;"><strong>The emphasis of this Notebook lies in analyzing the columns that will serve as features during the Modeling phase</strong></span>

<span style="color:#2929a3;font-size:20px;">Import Libraries</span>

In [55]:
import pandas as pd
import numpy as np
import plotly.express as px
from sklearn.preprocessing import MultiLabelBinarizer
import plotly.graph_objects as go

<span style="color:#2929a3;font-size:20px;">Read Dataset </span>

In [56]:
original_data = pd.read_csv('Data/Features.csv')

<span style="color:#2929a3;font-size:20px;">Show Sample of The Dataset </span>

In [57]:
original_data.sample(5)

DevType  \
40146                              Developer, mobile   
46753                          Developer, full-stack   
54860  Developer, desktop or enterprise applications   
81722                    Developer, game or graphics   
58274                                            NaN   

                                  LanguageHaveWorkedWith  \
40146                                              Swift   
46753                            HTML/CSS;JavaScript;SQL   
54860                           C#;C++;PowerShell;Python   
81722  Groovy;HTML/CSS;Java;JavaScript;Kotlin;TypeScript   
58274                HTML/CSS;Java;JavaScript;TypeScript   

             DatabaseHaveWorkedWith                  PlatformHaveWorkedWith  \
40146                  MySQL;SQLite      Amazon Web Services (AWS);Firebase   
46753                 Elasticsearch  Amazon Web Services (AWS);Google Cloud   
54860                           NaN                                     NaN   
81722  Dynamodb;PostgreSQL;Supabase        Amazon Web Services (AWS);Vercel   
58274             Oracle;PostgreSQL               Amazon Web Services (AWS)   

              WebframeHaveWorkedWith  \
40146                            NaN   
46753          Express;Node.js;React   
54860                            NaN   
81722  Express;Next.js;Node.js;React   
58274          NestJS;Node.js;Vue.js   

                             MiscTechHaveWorkedWith  \
40146                                       SwiftUI   
46753                         RabbitMQ;React Native   
54860  .NET (5+) ;.NET Framework (1.0 - 4.8);MFC;Qt   
81722                          Electron;Ktor;OpenGL   
58274                                  Apache Kafka   

                                 ToolsTechHaveWorkedWith  \
40146                                                NaN   
46753                        Docker;Homebrew;npm;Webpack   
54860  Chocolatey;MSBuild;MSVC;NuGet;Pacman;Pip;Visua...   
81722        Ant;Chocolatey;Gradle;npm;Vite;Webpack;Yarn   
58274                                           npm;Vite   

                            NEWCollabToolsHaveWorkedWith  
40146                                              Xcode  
46753                                           WebStorm  
54860  Eclipse;Nano;Notepad++;Qt Creator;Rad Studio (...  
81722  Eclipse;IntelliJ IDEA;Nano;Visual Studio Code;...  
58274                             IntelliJ IDEA;WebStorm

<span style="color:#2929a3;font-size:20px;">Show Shape of Dataset </span>

In [58]:
original_data.shape

(89184, 8)

<span style="color:#2929a3;font-size:20px;">Check Duplication </span>

In [59]:
original_data.duplicated().sum()

1961

In [60]:
original_data.drop_duplicates(inplace=True)

In [61]:
original_data.reset_index(drop=True, inplace=True)

<span style="color:#2929a3;font-size:20px;">Check Null Values</span>

In [62]:
original_data.isnull().sum()

DevType                         10812
LanguageHaveWorkedWith            409
DatabaseHaveWorkedWith          13850
PlatformHaveWorkedWith          23620
WebframeHaveWorkedWith          20329
MiscTechHaveWorkedWith          30255
ToolsTechHaveWorkedWith          9406
NEWCollabToolsHaveWorkedWith     1639
dtype: int64

<span style="color:#2929a3;font-size:20px;">Read Info of The Dataset</span>

In [63]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87223 entries, 0 to 87222
Data columns (total 8 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   DevType                       76411 non-null  object
 1   LanguageHaveWorkedWith        86814 non-null  object
 2   DatabaseHaveWorkedWith        73373 non-null  object
 3   PlatformHaveWorkedWith        63603 non-null  object
 4   WebframeHaveWorkedWith        66894 non-null  object
 5   MiscTechHaveWorkedWith        56968 non-null  object
 6   ToolsTechHaveWorkedWith       77817 non-null  object
 7   NEWCollabToolsHaveWorkedWith  85584 non-null  object
dtypes: object(8)
memory usage: 5.3+ MB


<span style="color:#2929a3;font-size:20px;">Drop Rows that has Null values in DevType(Target)</span>

In [64]:
target_drop_idx = original_data[original_data['DevType'].isnull() == 1].index

In [65]:
original_data.drop(target_drop_idx, axis=0, inplace=True)

In [66]:
original_data.reset_index(drop=True, inplace=True)

<span style="color:#2929a3;font-size:20px;">Apply Split_Answer function</span>

In [67]:
## Create Split_Answers Funcion
def Split_Answers(answer):
    if isinstance(answer, str):
        return answer.split(';')
    else:
        ## if answer is null function returns empty list
        return []

In [68]:
## Apply Split_Answer Funcion to each column in the dataset
data = pd.DataFrame()
for col in original_data.drop('DevType', axis=1).columns:
    data[col] = original_data[col].apply(Split_Answers)
data['DevType'] = original_data['DevType']

In [69]:
data.sample(5)

LanguageHaveWorkedWith  \
59488                                               [C#]   
51019                                     [C#, F#, Rust]   
46366  [Bash/Shell (all shells), HTML/CSS, Java, Java...   
47792                      [HTML/CSS, Java, Python, SQL]   
8221                         [HTML/CSS, JavaScript, PHP]   

                   DatabaseHaveWorkedWith  \
59488                             [MySQL]   
51019  [Microsoft SQL Server, PostgreSQL]   
46366               [MongoDB, PostgreSQL]   
47792           [MariaDB, Oracle, SQLite]   
8221                                   []   

                          PlatformHaveWorkedWith  \
59488  [Amazon Web Services (AWS), Google Cloud]   
51019   [Digital Ocean, Microsoft Azure, Render]   
46366                [Amazon Web Services (AWS)]   
47792                                         []   
8221                                          []   

                                  WebframeHaveWorkedWith  \
59488                                        [WordPress]   
51019                    [Angular, ASP.NET CORE, Svelte]   
46366  [Express, jQuery, Node.js, React, Spring Boot,...   
47792                                            [Flask]   
8221                                            [jQuery]   

      MiscTechHaveWorkedWith  \
59488                     []   
51019           [.NET (5+) ]   
46366     [Spring Framework]   
47792                     []   
8221               [Xamarin]   

                                 ToolsTechHaveWorkedWith  \
59488                 [Unity 3D, Visual Studio Solution]   
51019                                     [Cargo, NuGet]   
46366  [Composer, Docker, Maven (build tool), npm, We...   
47792                  [Docker, Maven (build tool), Pip]   
8221                                                  []   

                            NEWCollabToolsHaveWorkedWith  \
59488     [Notepad++, Visual Studio, Visual Studio Code]   
51019  [CLion, Goland, Visual Studio, Visual Studio C...   
46366  [IntelliJ IDEA, Notepad++, Vim, Visual Studio ...   
47792                 [IntelliJ IDEA, Sublime Text, Vim]   
8221                          [Atom, Visual Studio Code]   

                           DevType  
59488  Developer, game or graphics  
51019            DevOps specialist  
46366        Developer, full-stack  
47792        Developer, full-stack  
8221         Developer, full-stack

<span style="color:#2929a3;font-size:20px;">Apply MultiLabel Binarizer to each column</span>

In [70]:
encoded_dfs = {}
for col in data.drop('DevType', axis=1).columns:
    mlb = MultiLabelBinarizer()
    mlb_df = pd.DataFrame(mlb.fit_transform(data[col]), columns=mlb.classes_)
    encoded_dfs[col] = mlb_df

In [71]:
## Merge encoded_dfs in one dataframe
df = pd.concat(encoded_dfs, axis=1)
df['DevType'] = data['DevType']

In [72]:
df.sample(5) 

LanguageHaveWorkedWith                                                  \
                         APL Ada Apex Assembly Bash/Shell (all shells)  C C#   
72503                      0   0    0        0                       0  0  0   
20141                      0   0    0        0                       0  0  0   
14496                      0   0    0        0                       0  0  1   
43081                      0   0    0        0                       0  0  1   
33822                      0   0    0        0                       0  0  1   

                         ... NEWCollabToolsHaveWorkedWith                    \
      C++ Clojure Cobol  ...                 Sublime Text TextMate VSCodium   
72503   0       0     0  ...                            0        0        0   
20141   0       0     0  ...                            0        0        0   
14496   0       0     0  ...                            0        0        0   
43081   0       0     0  ...                            1        0        0   
33822   0       0     0  ...                            0        0        0   

                                                                 \
      Vim Visual Studio Visual Studio Code WebStorm Xcode condo   
72503   0             0                  1        0     0     0   
20141   0             0                  0        0     0     0   
14496   0             0                  1        0     0     0   
43081   0             0                  0        0     1     0   
33822   0             1                  1        0     0     0   

                                             DevType  
                                                      
72503                           Developer, front-end  
20141                          Developer, full-stack  
14496                          Developer, full-stack  
43081                          Security professional  
33822  Developer, desktop or enterprise applications  

[5 rows x 271 columns]

In [73]:
## Display DevType Frequency
df['DevType'].value_counts().reset_index()

DevType  count
0                           Developer, full-stack  25689
1                             Developer, back-end  13705
2                            Developer, front-end   5018
3   Developer, desktop or enterprise applications   3848
4                         Other (please specify):   3025
5                               Developer, mobile   2520
6                             Engineering manager   2025
7                                         Student   1966
8     Developer, embedded applications or devices   1837
9   Data scientist or machine learning specialist   1585
10                              DevOps specialist   1387
11                    Research & Development role   1348
12                            Academic researcher   1335
13           Senior Executive (C-Suite, VP, etc.)   1330
14                                 Engineer, data   1243
15                  Cloud infrastructure engineer   1036
16                    Developer, game or graphics    861
17                       Data or business analyst    829
18                           System administrator    735
19                                Project manager    585
20                          Developer, QA or test    582
21                          Security professional    472
22                                Product manager    443
23                     Engineer, site reliability    427
24                                       Educator    413
25                                      Scientist    347
26                           Developer Experience    326
27                                     Blockchain    322
28                              Hardware Engineer    283
29                                       Designer    273
30                         Database administrator    255
31                             Developer Advocate    212
32                Marketing or sales professional    149

<span style="color:#2929a3;font-size:20px;">Display tree map for all skills  <span> 

In [74]:
skills_freq = df.drop('DevType', axis=1).sum().reset_index()
skills_freq.columns = ['Group', 'Skill', 'Freq']
#skills_freq

In [75]:
fig_tree = px.treemap(skills_freq, path=['Group', 'Skill'],
                      values='Freq',
                      color='Freq', 
                      color_continuous_scale='deep')
fig_tree.show()

<span style="color:#2929a3;font-size:20px;">Create Jobs & Skills Heatmap<span> 

In [76]:
sorted_jobs = df['DevType'].value_counts(ascending=True).index.tolist()

In [77]:
sorted_skills = df.drop('DevType', axis=1).sum().sort_values(ascending=False).droplevel(level=0).index.tolist()

In [78]:
# dict for percentage of each skill exist in each skill column for each job
skills_jobs = {}

# iterate for each job
for job in sorted_jobs:
    skills_job = {} # dict for percentage of each skill exist in each skill column for one job
    
    # iterate for each skill column 
    for skill in data.drop('DevType', axis=1).columns:
        skill_job = df[df['DevType'] == job][skill].mean() * 100 
        skills_job[skill] = skill_job
        
    # concat dict of each skill column percentage for one job
    skills_job = pd.concat(skills_job)
    skills_jobs[job] = skills_job
    
# concat dict of each skill percentage of each job
skills_jobs = pd.concat(skills_jobs, axis=1)

In [79]:
skills_jobs = skills_jobs.reset_index(level=0, drop=True)
skills_jobs = skills_jobs.loc[sorted_skills]
skills_jobs = skills_jobs.T

In [80]:
skills_jobs

Visual Studio Code  JavaScript  \
Marketing or sales professional                         57.718121   56.375839   
Developer Advocate                                      73.584906   63.679245   
Database administrator                                  51.764706   29.803922   
Designer                                                59.706960   63.003663   
Hardware Engineer                                       59.363958   25.795053   
Blockchain                                              68.322981   62.422360   
Developer Experience                                    73.312883   55.214724   
Scientist                                               52.161383   27.953890   
Educator                                                65.859564   58.111380   
Engineer, site reliability                              71.896956   49.882904   
Product manager                                         67.494357   66.591422   
Security professional                                   68.855932   57.415254   
Developer, QA or test                                   72.680412   53.780069   
Project manager                                         67.008547   61.709402   
System administrator                                    60.544218   52.653061   
Data or business analyst                                63.329312   40.892642   
Developer, game or graphics                             70.383275   43.786295   
Cloud infrastructure engineer                           73.938224   50.579151   
Engineer, data                                          69.509252   35.961384   
Senior Executive (C-Suite, VP, etc.)                    69.548872   72.406015   
Academic researcher                                     59.176030   34.681648   
Research & Development role                             65.727003   40.949555   
DevOps specialist                                       76.063446   52.631579   
Data scientist or machine learning specialist           73.186120   31.419558   
Developer, embedded applications or devices             69.188895   30.157866   
Student                                                 80.213632   68.260427   
Engineering manager                                     72.444444   66.123457   
Developer, mobile                                       62.777778   37.063492   
Other (please specify):                                 67.834711   55.900826   
Developer, desktop or enterprise applications           61.564449   41.995842   
Developer, front-end                                    85.093663   91.988840   
Developer, back-end                                     66.661802   50.251733   
Developer, full-stack                                   79.458134   85.565806   

                                                HTML/CSS        SQL  \
Marketing or sales professional                55.704698  40.939597   
Developer Advocate                             56.603774  46.698113   
Database administrator                         31.764706  82.745098   
Designer                                       64.468864  30.769231   
Hardware Engineer                              27.915194  22.261484   
Blockchain                                     38.509317  30.434783   
Developer Experience                           38.343558  36.503067   
Scientist                                      25.360231  28.242075   
Educator                                       56.416465  46.004843   
Engineer, site reliability                     40.046838  44.496487   
Product manager                                53.724605  54.853273   
Security professional                          44.915254  45.762712   
Developer, QA or test                          38.144330  39.003436   
Project manager                                55.555556  56.410256   
System administrator                           51.700680  53.061224   
Data or business analyst                       40.289505  72.858866   
Developer, game or graphics                    33.797909  19.512195   
Cloud infrastr

In [81]:
fig = go.Figure(data=go.Heatmap(z=skills_jobs, x = skills_jobs.columns , y = skills_jobs.index))
fig.update_layout(width=5000, height=800)
fig.show()
fig.write_html('Skills_Jobs.html')

<span style="color:#2929a3;font-size:20px;">Create Scaled Jobs & Skills Heatmap<span> 

In [82]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_skills_jobs = pd.DataFrame(scaler.fit_transform(skills_jobs), columns=skills_jobs.columns, index=skills_jobs.index)

In [83]:
fig = go.Figure(data=go.Heatmap(z=scaled_skills_jobs, x = scaled_skills_jobs.columns , y = scaled_skills_jobs.index))
fig.update_layout(width=5000, height=800)
fig.show()
fig.write_html('Scaled_Skills_Jobs.html')

## From the previous visualizations we noticed that scaling the data makes the value of each skill more expressive to each job

 
- For example before scaling the percentage of Keras in Data Science is 34.3 out of 100 but after scaling the value becomes 5.02 out of 6 which is more expressive

- Another example , before scaling Unity 3D in Game/Graphics Developing is 44.7 out of 100 but after scaling the value becomes 5.4 out of 6 which is also more expressive

- Another example , before scaling Visual Studio Code in Database Administration is 51.7 out of 100 but after scaling the value becomes -2.1 out of 6 which is also more expressive .. etc

<span style="color:#2929a3;font-size:20px;">Jop Profiles<span> 

In [84]:
#job = np.random.choice(sorted_jobs)
job = 'Data scientist or machine learning specialist'
#job = 'Marketing or sales professional'
#job = 'Developer, back-end'
#job = 'Developer, front-end'
#job = 'Developer, full-stack'
#job = 'Data or business analyst'
#job = 'Database administrator'

In [85]:
job

'Data scientist or machine learning specialist'

In [86]:
single_job_skills = pd.concat([skills_jobs.loc[job], scaled_skills_jobs.loc[job]], axis=1)
single_job_skills.columns = ['Percentage', 'Scaled_Value']
single_job_skills = single_job_skills.sort_values('Percentage')

In [87]:
single_job_skills.head()

Percentage  Scaled_Value
liblittletest         0.0     -0.606721
CUTE                  0.0     -0.537147
lest                  0.0     -0.855274
snitch                0.0     -0.788875
tunit                 0.0     -0.662249

In [88]:
## Comparing skills before and after scaling for random job
threshold = 25

single_job_skills = single_job_skills[single_job_skills['Percentage'] > threshold]

fig = px.bar(x=single_job_skills['Percentage'],
             y=single_job_skills.index,
             color=single_job_skills['Scaled_Value'],
             color_continuous_scale='orrd',
            range_color=[scaled_skills_jobs.min().min(), scaled_skills_jobs.max().max()])
fig.update_layout(title=job)
fig.show()

<span style="color:#2929a3;font-size:20px;">Display Dendogram for this Data<span> 

In [89]:
## Dict for replace long columns names
dict_Devtype = {'Senior Executive (C-Suite, VP, etc.)':'Senior Executive',
                'Developer, back-end':'Back-End',
                'Developer, front-end':'Front-End',
                'Developer, full-stack':'Full-Stack',
                'Developer, desktop or enterprise applications':'Desktop Developer',
                'Developer, QA or test':'Tester',
                'Data scientist or machine learning specialist':'Data Scientist',
                'Data or business analyst':'Data Analyst',
                'Developer, embedded applications or devices':'Embedded Developer',
                'Marketing or sales professional':'Marketing/Sales',  
}

short_labels = scaled_skills_jobs.index.to_series().replace(dict_Devtype).tolist()

import scipy.cluster.hierarchy as sch
import plotly.figure_factory as ff

fig_dendo = ff.create_dendrogram(scaled_skills_jobs, labels=short_labels, orientation='left')
fig_dendo.update_layout(width=1000, height=600)
fig_dendo.show()

<span style="color:#2929a3;font-size:20px;">Drop rows that has illogical Number of skills from original dataframe (Data) <span> 

In [90]:
data.head()

LanguageHaveWorkedWith  \
0                     [HTML/CSS, JavaScript, Python]   
1                      [Bash/Shell (all shells), Go]   
2  [Bash/Shell (all shells), HTML/CSS, JavaScript...   
3                 [HTML/CSS, JavaScript, TypeScript]   
4  [Bash/Shell (all shells), HTML/CSS, JavaScript...   

                           DatabaseHaveWorkedWith  \
0                                      [Supabase]   
1                                              []   
2                             [PostgreSQL, Redis]   
3  [BigQuery, Elasticsearch, MongoDB, PostgreSQL]   
4  [BigQuery, Cloud Firestore, PostgreSQL, Redis]   

                              PlatformHaveWorkedWith  \
0       [Amazon Web Services (AWS), Netlify, Vercel]   
1  [Amazon Web Services (AWS), Google Cloud, Open...   
2                               [Cloudflare, Heroku]   
3  [Amazon Web Services (AWS), Firebase, Heroku, ...   
4  [Amazon Web Services (AWS), Cloudflare, Google...   

                              WebframeHaveWorkedWith  \
0                    [Next.js, React, Remix, Vue.js]   
1                                                 []   
2  [Node.js, React, Ruby on Rails, Vue.js, WordPr...   
3  [Express, Gatsby, NestJS, Next.js, Node.js, Re...   
4                [Angular, Express, NestJS, Node.js]   

            MiscTechHaveWorkedWith  \
0  [Electron, React Native, Tauri]   
1                               []   
2                               []   
3                               []   
4                               []   

                             ToolsTechHaveWorkedWith  \
0  [Docker, Kubernetes, npm, Pip, Vite, Webpack, ...   
1             [Cargo, Docker, Kubernetes, Make, Nix]   
2               [Homebrew, npm, Vite, Webpack, Yarn]   
3                       [Docker, npm, Webpack, Yarn]   
4  [Docker, Homebrew, Kubernetes, npm, pnpm, Terr...   

                        NEWCollabToolsHaveWorkedWith  \
0                          [Vim, Visual Studio Code]   
1                                     [Emacs, Helix]   
2  [IntelliJ IDEA, Vim, Visual Studio Code, WebSt...   
3                          [Vim, Visual Studio Code]   
4                                    [Helix, Neovim]   

                                DevType  
0  Senior Executive (C-Suite, VP, etc.)  
1                   Developer, back-end  
2                  Developer, front-end  
3                 Developer, full-stack  
4                   Developer, back-end

In [91]:
def Count_Total_Skills(row):
    return sum(len(List) for List in row)

In [92]:
Total_Skills_Count = data[data.columns[1:]].apply(Count_Total_Skills, axis=1)

In [93]:
px.histogram(Total_Skills_Count, title='Total Skills Count')

In [94]:
## Drop rows that has No of Skills more thar 40 or less than 4
drop_skills_idx = Total_Skills_Count[(Total_Skills_Count > 40) | (Total_Skills_Count < 4)].index

In [95]:
## Drop from both data and original data 
data.drop(drop_skills_idx, axis=0, inplace=True)
original_data.drop(drop_skills_idx, axis=0, inplace=True)

# reset index of both data and original data
data.reset_index(drop=True, inplace=True)
original_data.reset_index(drop=True, inplace=True)

<span style="color:#2929a3;font-size:20px;">Drop rows that has general jobs or not important jobs<span> 

In [96]:
excluded_jobs = [
    'Other (please specify):',
    'Student',
    'Designer',
    'Educator',
    'Marketing or sales professional',
    'Engineering manager',
    'Senior Executive (C-Suite, VP, etc.)',
    'Product manager',
    'Project manager',
    'Developer Advocate',
    'Developer Experience',
    'Scientist',
    'Engineer, site reliability',
    'Academic researcher',
    'Research & Development role']

In [97]:
drop_DevType_idx = data[data['DevType'].isin(excluded_jobs)].index

In [98]:
## Drop from both data and original data
data.drop(drop_DevType_idx, axis=0, inplace=True)
original_data.drop(drop_DevType_idx, axis=0, inplace=True)

## Reset index of both data and original data
data.reset_index(drop=True, inplace=True)
original_data.reset_index(drop=True, inplace=True)

<span style="color:#2929a3;font-size:20px;">Merge Close Jobs<span> 

In [99]:
## Merge Embedded System Developer With Hardware Developer
close_job1_idx = data[data['DevType'].isin(['Developer, embedded applications or devices', 'Hardware Engineer'])].index

In [100]:
data['DevType'].loc[close_job1_idx] = "Developer, Hardware/Embedded Systems"
original_data['DevType'].loc[close_job1_idx] = "Developer, Hardware/Embedded Systems"

In [101]:
## Merge System Adminstration with Database Adminstration
close_job2_idx = data[data['DevType'].isin(['Database administrator','System administrator'])].index

In [102]:
data['DevType'].loc[close_job2_idx] = "Database/System Admininstrator"
original_data['DevType'].loc[close_job2_idx] = "Database/System Admininstrator"

In [103]:
## Merge Front-End , Back-End and Full-Stack as Web Developer
close_job3_idx = data[data['DevType'].isin(['Developer, back-end','Developer, front-end', 'Developer, full-stack'])].index

In [104]:
data['DevType'].loc[close_job3_idx] = "Web Developer"
original_data['DevType'].loc[close_job3_idx] = "Web Developer"

In [105]:
original_data['DevType'].loc[close_job1_idx]

40       Developer, Hardware/Embedded Systems
519      Developer, Hardware/Embedded Systems
655      Developer, Hardware/Embedded Systems
732      Developer, Hardware/Embedded Systems
1008     Developer, Hardware/Embedded Systems
                         ...                 
38036    Developer, Hardware/Embedded Systems
38076    Developer, Hardware/Embedded Systems
38132    Developer, Hardware/Embedded Systems
38228    Developer, Hardware/Embedded Systems
38316    Developer, Hardware/Embedded Systems
Name: DevType, Length: 263, dtype: object

In [106]:
## Check Final Shape
data.shape

(38545, 8)

In [109]:
data

LanguageHaveWorkedWith  \
0                          [Bash/Shell (all shells), Go]   
1      [Bash/Shell (all shells), HTML/CSS, JavaScript...   
2      [Bash/Shell (all shells), HTML/CSS, JavaScript...   
3                                 [C, C++, Python, Rust]   
4                               [Java, Perl, TypeScript]   
...                                                  ...   
38540  [Bash/Shell (all shells), HTML/CSS, JavaScript...   
38541                 [C#, HTML/CSS, JavaScript, R, SQL]   
38542                                     [Java, Kotlin]   
38543                           [JavaScript, TypeScript]   
38544                          [Dart, Java, Python, SQL]   

                               DatabaseHaveWorkedWith  \
0                                                  []   
1                                 [PostgreSQL, Redis]   
2      [BigQuery, Cloud Firestore, PostgreSQL, Redis]   
3                                             [Redis]   
4                                             [MySQL]   
...                                               ...   
38540                    [Oracle, PostgreSQL, SQLite]   
38541                   [Microsoft SQL Server, MySQL]   
38542                    [H2, MongoDB, MySQL, SQLite]   
38543                                      [Dynamodb]   
38544                                    [PostgreSQL]   

                                  PlatformHaveWorkedWith  \
0      [Amazon Web Services (AWS), Google Cloud, Open...   
1                                   [Cloudflare, Heroku]   
2      [Amazon Web Services (AWS), Cloudflare, Google...   
3                                                     []   
4                                                     []   
...                                                  ...   
38540                                           [VMware]   
38541                                                 []   
38542                                           [VMware]   
38543                        [Amazon Web Services (AWS)]   
38544                                                 []   

                                  WebframeHaveWorkedWith  \
0                                                     []   
1      [Node.js, React, Ruby on Rails, Vue.js, WordPr...   
2                    [Angular, Express, NestJS, Node.js]   
3                                                     []   
4                              [Fastify, Node.js, React]   
...                                                  ...   
38540                         [Express, jQuery, Node.js]   
38541                          [ASP.NET CORE, WordPress]   
38542                                      [Spring Boot]   
38543                    [Express, Node.js, Spring Boot]   
38544                                      [Spring Boot]   

                                  MiscTechHaveWorkedWith  \
0                                                     []   
1                                                     []   
2                                                     []   
3                                                     []   
4                                     [Spring Framework]   
...                                                  ...   
38540                                                 []   
38541                                          [Cordova]   
38542                 [Opencv, OpenGL, Spring Framework]   
38543                                     [Apache Kafka]   
38544  [Flutter, Opencv, RabbitMQ, Spring Framework, ...   

                                 ToolsTechHaveWorkedWith  \
0                 [Cargo, Docker, Kubernetes, Make, Nix]   
1                   [Homebrew, npm, Vite, Webpack, Yarn]   
2      [Docker, Homebrew, Kubernetes, npm, pnpm, Terr...   
3                  [Cargo, CMake, Docker, GNU GCC, Make]   
4                                     [Kubernetes, Yarn]   
...                                                  ...   
38540            [APT, GNU GCC, Make, npm, Pi

In [107]:
original_data.shape

(38545, 8)

In [108]:
## Dump Final Features 
data.to_pickle('Data/Cleaned_Features.pkl')
original_data.to_pickle('Data/Cleaned_original_data.pkl')

In [110]:
original_data

DevType  \
0              Web Developer   
1              Web Developer   
2              Web Developer   
3      Developer, QA or test   
4              Web Developer   
...                      ...   
38540          Web Developer   
38541          Web Developer   
38542      Developer, mobile   
38543          Web Developer   
38544          Web Developer   

                                  LanguageHaveWorkedWith  \
0                             Bash/Shell (all shells);Go   
1      Bash/Shell (all shells);HTML/CSS;JavaScript;PH...   
2      Bash/Shell (all shells);HTML/CSS;JavaScript;Ru...   
3                                      C;C++;Python;Rust   
4                                   Java;Perl;TypeScript   
...                                                  ...   
38540  Bash/Shell (all shells);HTML/CSS;JavaScript;Pe...   
38541                       C#;HTML/CSS;JavaScript;R;SQL   
38542                                        Java;Kotlin   
38543                              JavaScript;TypeScript   
38544                               Dart;Java;Python;SQL   

                          DatabaseHaveWorkedWith  \
0                                            NaN   
1                               PostgreSQL;Redis   
2      BigQuery;Cloud Firestore;PostgreSQL;Redis   
3                                          Redis   
4                                          MySQL   
...                                          ...   
38540                   Oracle;PostgreSQL;SQLite   
38541                 Microsoft SQL Server;MySQL   
38542                    H2;MongoDB;MySQL;SQLite   
38543                                   Dynamodb   
38544                                 PostgreSQL   

                                  PlatformHaveWorkedWith  \
0      Amazon Web Services (AWS);Google Cloud;OpenSta...   
1                                      Cloudflare;Heroku   
2      Amazon Web Services (AWS);Cloudflare;Google Cloud   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
38540                                             VMware   
38541                                                NaN   
38542                                             VMware   
38543                          Amazon Web Services (AWS)   
38544                                                NaN   

                             WebframeHaveWorkedWith  \
0                                               NaN   
1      Node.js;React;Ruby on Rails;Vue.js;WordPress   
2                    Angular;Express;NestJS;Node.js   
3                                               NaN   
4                             Fastify;Node.js;React   
...                                             ...   
38540                        Express;jQuery;Node.js   
38541                        ASP.NET CORE;WordPress   
38542                                   Spring Boot   
38543                   Express;Node.js;Spring Boot   
38544                                   Spring Boot   

                                  MiscTechHaveWorkedWith  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                       Spring Framework   
...                                                  ...   
38540                                                NaN   
38541                                            Cordova   
38542                     Opencv;OpenGL;Spring Framework   
38543                                       Apache Kafka   
38544  Flutter;Opencv;RabbitMQ;Spring Framework;Torch...   

                             ToolsTechHaveWorkedWith  \
0                   Cargo;Docker;Kubernetes;Make;Nix   
1                     Homebrew;npm;Vite;Webpack;Yarn   
2      Docker;Homeb